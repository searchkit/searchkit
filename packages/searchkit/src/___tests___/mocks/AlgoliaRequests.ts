export const SimpleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const nonDynamicFacetRequestOneFilter = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['actors:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const DisjunctiveExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['type:movie']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'type',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: 'shawshank',
      tagFilters: ''
    }
  }
]

export const NestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const TwoFiltersNestedQueryRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetFilters: [['user.first:Alice', 'user.first:John']],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      attributesToHighlight: [],
      attributesToRetrieve: [],
      attributesToSnippet: [],
      clickAnalytics: false,
      facets: 'user.first',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 1,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFilterNestedQueryRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetFilters: [],
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['user.price>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const FacetQueryExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facetName: 'type',
      facetQuery: 'm',
      facets: ['*'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]

export const NumericFiltersExampleRequest = [
  {
    indexName: 'imdb_movies',
    params: {
      facets: ['type', 'actors', 'imdbrating'],
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      maxValuesPerFacet: 10,
      numericFilters: ['imdbrating>=90'],
      page: 0,
      query: '',
      tagFilters: ''
    }
  },
  {
    indexName: 'imdb_movies',
    params: {
      analytics: false,
      clickAnalytics: false,
      facets: 'imdbrating',
      highlightPostTag: '</em>',
      highlightPreTag: '<em>',
      hitsPerPage: 0,
      maxValuesPerFacet: 10,
      page: 0,
      query: ''
    }
  }
]

export const NestedFacetQueryExampleRequest = [
  {
    indexName: 'my-index-000001',
    params: {
      facetName: 'user.first',
      facetQuery: 'har',
      facets: ['user.first', 'user.price'],
      highlightPostTag: '</ais-highlight-0000000000>',
      highlightPreTag: '<ais-highlight-0000000000>',
      maxFacetHits: 10,
      maxValuesPerFacet: 10,
      page: 0,
      query: '',
      tagFilters: ''
    }
  }
]
