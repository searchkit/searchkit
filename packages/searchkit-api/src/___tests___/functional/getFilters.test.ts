import Client, { AlgoliaMultipleQueriesQuery } from '../../'
import nock from 'nock'
import { DisjunctiveExampleRequest } from '../mocks/AlgoliaRequests'
import { HitsResponseWithFacetFilter } from '../mocks/ElasticsearchResponses'

describe('Add additional base filters to search', () => {
  it('call with one filter and query applied', async () => {
    const client = Client({
      connection: {
        host: 'https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243',
        apiKey: 'a2Rha1VJTUJMcGU4ajA3Tm9fZ0Y6MjAzX2pLbURTXy1hNm9SUGZGRlhJdw=='
      },
      search_settings: {
        highlight_attributes: ['title', 'actors'],
        search_attributes: ['title', 'actors', 'query'],
        result_attributes: ['title', 'actors', 'query'],
        facet_attributes: [
          'type',
          { field: 'actors.keyword', attribute: 'actors', type: 'string' },
          'rated'
        ]
      }
    })

    nock('https://commerce-demo.es.us-east4.gcp.elastic-cloud.com:9243')
      .post('/_msearch', (requestBody: any) => {
        expect(requestBody).toMatchSnapshot('ES Request')
        const x = JSON.parse(requestBody.split('\n')[1]).query
        // has the base filter applied in addition to the facet filter
        expect(x).toMatchInlineSnapshot(`
          {
            "bool": {
              "filter": [
                {
                  "bool": {
                    "should": [
                      {
                        "term": {
                          "type": "movie",
                        },
                      },
                    ],
                  },
                },
                {
                  "bool": {
                    "must": {
                      "range": {
                        "imdbrating": {
                          "gte": 1,
                        },
                      },
                    },
                  },
                },
              ],
              "must": {
                "function_score": {
                  "functions": [],
                  "query": {
                    "pinned": {
                      "ids": [],
                      "organic": {
                        "combined_fields": {
                          "fields": [
                            "title",
                            "actors",
                            "query",
                          ],
                          "query": "shawshank",
                        },
                      },
                    },
                  },
                },
              },
            },
          }
        `)
        return true
      })
      .reply(200, HitsResponseWithFacetFilter)

    const response = await client.handleInstantSearchRequests(
      DisjunctiveExampleRequest as AlgoliaMultipleQueriesQuery[],
      {
        getBaseFilters: () => {
          return [
            {
              bool: {
                must: {
                  range: {
                    imdbrating: {
                      gte: 1
                    }
                  }
                }
              }
            }
          ]
        }
      }
    )

    expect(response).toMatchSnapshot()
  })
})
