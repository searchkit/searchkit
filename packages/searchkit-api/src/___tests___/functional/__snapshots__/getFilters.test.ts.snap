// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add additional base filters to search call with one filter and query applied 2`] = `
{
  "results": [
    {
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "actors": {
          "Bob Gunton": 1,
          "Morgan Freeman": 1,
          "Tim Robbins": 1,
          "William Sadler": 1,
        },
        "rated": {
          "R": 1,
        },
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "actors": [
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Tim Robbins",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Morgan Freeman",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Bob Gunton",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "William Sadler",
              },
            ],
            "title": [
              {
                "fullyHighlighted": false,
                "matchLevel": "full",
                "matchedWords": [
                  "Shawshank",
                ],
                "value": "The <ais-highlight-0000000000>Shawshank</ais-highlight-0000000000> Redemption",
              },
            ],
          },
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 20,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "facetFilters=&facets=*&highlightPostTag=%3C%2Fais-highlight-0000000000%3E&highlightPreTag=%3Cais-highlight-0000000000%3E&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "type",
              "actors",
              "rated",
            ],
          },
          "values": {
            "actors": {
              "sortRemainingBy": "count",
            },
            "rated": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
    {
      "exhaustive": {
        "facetsCount": true,
        "nbHits": true,
        "typo": true,
      },
      "exhaustiveFacetsCount": true,
      "exhaustiveNbHits": true,
      "exhaustiveTypo": true,
      "facets": {
        "type": {
          "movie": 1,
        },
      },
      "facets_stats": {},
      "hits": [
        {
          "_highlightResult": {
            "actors": [
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Tim Robbins",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Morgan Freeman",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "Bob Gunton",
              },
              {
                "matchLevel": "none",
                "matchedWords": [],
                "value": "William Sadler",
              },
            ],
            "title": [
              {
                "fullyHighlighted": false,
                "matchLevel": "full",
                "matchedWords": [
                  "Shawshank",
                ],
                "value": "The <ais-highlight-0000000000>Shawshank</ais-highlight-0000000000> Redemption",
              },
            ],
          },
          "actors": [
            "Tim Robbins",
            "Morgan Freeman",
            "Bob Gunton",
            "William Sadler",
          ],
          "objectID": "tt0111161",
          "title": "The Shawshank Redemption",
        },
      ],
      "hitsPerPage": 1,
      "index": "imdb_movies",
      "nbHits": 1,
      "nbPages": 1,
      "page": 0,
      "params": "analytics=false&clickAnalytics=false&facets=type&highlightPostTag=%3C%2Fais-highlight-0000000000%3E&highlightPreTag=%3Cais-highlight-0000000000%3E&hitsPerPage=1&maxValuesPerFacet=10&page=0&query=shawshank&tagFilters=",
      "processingTimeMS": 2,
      "query": "shawshank",
      "renderingContent": {
        "facetOrdering": {
          "facets": {
            "order": [
              "type",
              "actors",
              "rated",
            ],
          },
          "values": {
            "actors": {
              "sortRemainingBy": "count",
            },
            "rated": {
              "sortRemainingBy": "count",
            },
            "type": {
              "sortRemainingBy": "count",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`Add additional base filters to search call with one filter and query applied: ES Request 1`] = `
"{"index":"imdb_movies"}
{"aggs":{"type":{"terms":{"field":"type","size":10}},"actors":{"terms":{"field":"actors.keyword","size":10}},"rated":{"terms":{"field":"rated","size":10}}},"query":{"bool":{"filter":[{"bool":{"should":[{"term":{"type":"movie"}}]}},{"bool":{"must":{"range":{"imdbrating":{"gte":1}}}}}],"must":{"combined_fields":{"query":"shawshank","fields":["title","actors","query"]}}}},"size":20,"from":0,"_source":{"includes":["title","actors","query"]},"highlight":{"pre_tags":["<ais-highlight-0000000000>"],"post_tags":["</ais-highlight-0000000000>"],"fields":{"title":{},"actors":{}}}}
{"index":"imdb_movies"}
{"aggs":{"type":{"terms":{"field":"type","size":10}}},"query":{"bool":{"filter":[{"bool":{"must":{"range":{"imdbrating":{"gte":1}}}}}],"must":{"combined_fields":{"query":"shawshank","fields":["title","actors","query"]}}}},"size":1,"from":0,"_source":{"includes":["title","actors","query"]},"highlight":{"pre_tags":["<ais-highlight-0000000000>"],"post_tags":["</ais-highlight-0000000000>"],"fields":{"title":{},"actors":{}}}}
"
`;
